### DISCUSSION TITLE
License discussion

### DISCUSSION THREAD

**Discussion by SorraTheOrc** [hlky/stable-diffusion-webui#235] [General]
**Describe the bug**
Without a clear license on the code it is unsafe to use this in any environment. I see your response to a previous request on https://github.com/hlky/stable-diffusion-webui/issues?q=is%3Aissue+license+is%3Aclosed but it isn't an appropriate answer. An open source license gives people the rights they need to use and enhance your code, it does not give any warranties or support expectation.

As things stand your code cannot be built upon by others, like me, for any purpose.

Please consider putting a permissive open source license on this code such as https://choosealicense.com/licenses/mit/

One limitation you may have is that you derive your code from another persons. If that persons code is also unlicensed then we need to get them to put a license on their code. Without it you are (technically) doing something you do not have the rights to do.

**Comment by SorraTheOrc** [comment]
I made a request on the upstream project too https://github.com/AUTOMATIC1111/stable-diffusion-webui/issues/24

**Comment by hlky** [comment]
It is unlikely I will add a license that allows commercial usage such as MIT.

In my opinion, with a project like this, it is unfair to allow others to take this code and use it for commercial gain.

If you can tell me what your intended use is that will help me make a decision.

**Comment by SorraTheOrc** [comment]
My intention is to help you make it better by contributing to it as open source. I don't have any plans to use it commercially but who knows what I would want to do in the future. I would want people to use it commercially now.

You are free to do what you want with your own code. It's a shame, but it's your choice.

Note, free GitHub accounts are ONLY for open source projects - https://github.com/pricing

**Reply by JohannesGaessler** [reply]
What specific part of the linked page are you referring to?
I am not a legal expert but my understanding is that this repository is open source, it is simply not licensed.
The legal consequence is that copyright holders such as hlky and me have reserved all rights to arbitrarily restrict its use by other people.

In any case, if Github were to enforce a rule regarding licensing that goes against the wishes of the project developers I would prefer to simply switch hosting providers.
Personally I favor licensing the project under AGPL.

**Comment by hlky** [comment]
Can you suggest a license that does restrict commercial usage, but doesn't have any 'weird' things and won't cause issues if I do decide to relicense in the future?

Honestly imo any commercial project would be better off starting from scratch.

**Comment by JoshuaKimsey** [comment]
IDK if this would help at all, but this is a decent post on what options there are for an open-source license that blocks most/all commercial usage: https://qr.ae/pvblfW

TL;DR, could consider using the AGPL license or go a more non-traditional (and not entirely kosher) route and license it under **CC BY-NC-SA 2.0**. The latter might be the better bet for now, but I'll leave that up to you. Otherwise, the GPL might be a decent option for the project, or maybe even the Apache if you don't want the hard-line copyleft aspect of the GPL.

**Reply by SorraTheOrc** [reply]
There is no such thing as an open source license that limits commercial use. The open source definition does not allow for any exclusionary access to the code, see https://opensource.org/osd

AGPL, nor Apache License limits commercial use. Apache is open source and was explicitly created to enable allow commercial use (see history of Apache Software Foundation on Wikipedia). AGPL is not open source but is a free software license (which also does not limit commercial use). AGPL, like GPL requires derivatives to be released under the same license, permissive licensees like Apache do not. AGPL differs from GPL in that it classes hosting as distribution and thus triggers the reciprocal licensing more often.

You should not use creative commons licenses for code, they are not intended for that purpose and do not adequately cover the rights to reuse, see https://creativecommons.org/faq/#can-i-apply-a-creative-commons-license-to-software

**Reply by JoshuaKimsey** [reply]
I'm aware about both of those options not being all that great. The standard OSS licence doesn't really fit with what hlky wants, and the CC license really shouldn't be used for software. That said, if nothing else I would use the CC BY-NC-SA license as a stop-gap license for the project.

It should be mentioned that the license that the project currently uses is the one passed down from the main Stable Diffusion repo, the Open-ML license. Which for the ML part goes in depth and stipulations, etc, but for the general software is basically a form of the MIT license. So as it stands right now, the project is using an MIT-like license.

**Comment by ghost** [comment]
it looks like the decision has been made unless someone wants to go change his/her mind:
https://github.com/AUTOMATIC1111/stable-diffusion-webui/issues/24

**Comment by SorraTheOrc** [comment]
The AGPL allows you to do pretty much anything you want with the cod, including commercial use. If you host it in a public server then you must make your modifications available under the same license (there's a few other requirements too, check the license). If you do not host it publicly or make no modifications then don't worry about it.

**Reply by hlky** [reply]
AGPL only applies to the webui related stuff, the original CompViz stable diffusion code is licensed under CreativeML Open RAIL M License.

Honestly @SorraTheOrc it's pretty obvious you intend to use this for commercial purposes and profit off the work of the contributors without intending on contributing back. Just start your own project from scratch and you won't have to worry about anything.

**Reply by hlky** [reply]
Alternatively, don't be a greedy capitalist and just enjoy Stable Diffusion in the way it was meant to be enjoyed, the way our Lord and Savior Emad Mostaque intended, open and free, not packaged up and sold.

**Reply by JoshuaKimsey** [reply]
Yeah, AGPL is pretty unfriendly to commercial usage as well. Most commercial ventures really don't like touching it, so it's probably a good fit for the webui part of the project.

**Reply by JohannesGaessler** [reply]
I currently have licensed https://github.com/JohannesGaessler/stable-diffusion-tools under Apache 2.0, but since it's on the edge where the FSF starts recommending including copyleft I would change the license to AGPL to match.

**Comment by Gerschel** [comment]
I notice that this discussion was in place when Automatic1111 had made a decision. Later he removed it. Does this put this project in a grey area, where it was done with good faith that there was a license at the time.  
I don't know the licensing conundrums that well. But what I'm looking to do, is contribute to a project that has some license, even if it's in a grey area.  
I had made an extension, but it is limited becuase it's an extension and not embedded in the main code. On a local development copy, I added the correct snippets into ui.py, and got it to work across all tabs. I decided to hold back contributing, because of the license issue.  

I would love to contribute, I've spent a considerable amount of time learning the base code and writing my code.  But I don't want someone to be able to steal it and make money from it while I struggle. I think a good contribution would look good on a portfolio, or something, it just feels right.
